ENH: Provisional support for NEP 18 (__array_function__ protocol)

neutrinoceros opened the pull request:
My initial motivation here was to add some unit representation to error messages when comparing two unyt_array instance via functions from `numpy.testing`, like so:

```python
import numpy as np
import unyt as un

a = [1, 2, 3] * un.cm
b = [1, 2, 3] * un.km
np.testing.assert_array_equal(a, b)
```
which yields, on master:

```
...
AssertionError:
Arrays are not equal

Mismatched elements: 3 / 3 (100%)
Max absolute difference: 299997.
Max relative difference: 0.99999
 x: unyt_array([1, 2, 3])
 y: unyt_array([1, 2, 3])
```
and on this branch
<details>
<summary>previous version</summary>

```
...
AssertionError:
Arrays are not equal

Mismatched elements: 3 / 3 (100%)
Max absolute difference: 299997. cm
Max relative difference: 0.99999 dimensionless
 x: unyt_array([1, 2, 3] cm)
 y: unyt_array([1, 2, 3] km)
```

</details>

edit:
```
AssertionError:
Arrays are not equal

Mismatched elements: 3 / 3 (100%)
Max absolute difference: 299997., units='cm'
Max relative difference: 0.99999, units='dimensionless'
 x: unyt_array([1, 2, 3], units='cm')
 y: unyt_array([1, 2, 3], units='km')
```

Incidentally, it turns out that fixing this necessitated a kick-off implementation of NEP 18, so this work laid the foundation to solve:
- [x] #69
- [x] #130
- [ ] #50 (most likely out of scope)

More broadly, implementing NEP 18 for any was the topic of #139.
Granted I need to take more time to check that I'm not going against the original intentions here. My current approach is that, since covering the whole numpy public API in one go seems like an gigantic task, I'm implementing `unyt_array.__array_function__` with a fallthrough condition: if a special case isn't implemented yet, just fallback to the raw numpy implem (which is currently the behaviour for _all_ functions subject to NEP 18). This way we can add support for more and more functions in a progressive way.
I'm going to set the bar low(ish) for now, and try and fix the already reported cases, as reported above, as a first step.


An important question to address is: what should be done in the general case where we don't have a custom implementation for an array function ?

1) transparently default to the raw numpy implementation without a warning (this is de facto what is done as of unyt 2.8.0, and will remain the case until NEP 18 isn't at least partially implemented).
2) same as 1. _but_ emit a warning (possibly, we could have a whitelist of functions that are known to be perfectly fine without a specialised implementation, for which no warning would be emitted) along the lines of 
```
UserWarning: calling `numpy.FUNC` on a unyt_array. Results may hold incorrect units. A future version of unyt will remove this warning, and possibly change the behaviour of this function to be dimensionally correct.
```
3) Error out

Option 1 this is the current implementation of this PR because I think it is the less disruptive or noisy one. My personal opinion is that it's probably okay to have incomplete support for NEP 18 for a couple release, as long as it is clearly stated in the release notes.

neutrinoceros commented:
I've gone the long way and started implemented a bunch of functions with obvious requirements.
TODO:

## inverse-like functions
- [x] `np.linalg.inv`
- [x] `np.linalg.tensorinv`
- [ ] `np.linalg.pinv` (done-ish, more testing needed for the rcond parameter)

## product-like functions
- [x] `np.dot`
- [ ] `np.inner`
- [ ] `np.outer`
- [ ] `np.matmul`

## histogram functions (as requested for #130)
- [x] `np.histogram`
- [x] `np.histogram2d`
- [x] `np.histogramdd`

neutrinoceros commented:
pre-commit.ci autofix

neutrinoceros commented:
@jzuhone do you think you'd be able to review this ? Is there someone else I could ask ?

jzuhone commented:
@neutrinoceros I can try to look at it next week

neutrinoceros commented:
@jzuhone I would like to propose we bundle this PR in the next release alongside the dask layer. IMO, each of these would justify a release on their own, so it's up to you :)

jzuhone commented:
@chrishavlin or @matthewturk could you look at this one?

ngoldbaum commented:
I'm a little concerned that the coverage of numpy functions isn't as extensive as users might like and that implementing support in a piecemeal fashion might lead to more breakage than users would prefer, since fundamentally we're changing the type of the outputs of numpy functions. If we're going to break that, it would be best to break them all at once instead of staggering that over many releases as support gets added in a piecemeal fashion.

unyt ships with a number of wrappers for numpy functions (`uconcatenate`, `ucross`, `udot`, etc). We should make sure we have replacements for all of those (since there's clearly already a need for those) and deprecate the wrappers. I know at least yt uses those wrappers.

If you want a hand with this I'm happy to help out.

neutrinoceros commented:
I agree that it'd be best to do it all in one clean go, and yes I would love a hand here. How should we proceed ? If you want to keep it all in a single PR feel free to push to my branch.

ngoldbaum commented:
Sure I’ll push to this branch

neutrinoceros commented:
@ngoldbaum do you know of a reliable way to generate the complete list of functions we should be covering ?

ngoldbaum commented:
Yes, see https://github.com/numpy/numpy/issues/15544#issuecomment-584453408 as well as other comments in that issue

ngoldbaum commented:
Can you make sure that the existing wrappers in unyt that this PR obsoletes (e.g. `udot`) are deprecated? You could probably rewrite them to be thin wrappers around the regular numpy functions now.

neutrinoceros commented:
good call, I completely missed that. I think it may be preferable _not_ to rewrite them at all in case there's _any_ unanticipated difference so users have a chance to report and still get something that works for them while we fix the replacement, but I'll definitely add those deprecation warnings.

neutrinoceros commented:
I rebased the branch but didn't write the history. I added two commit, the first of which corresponds to #288, and the second one adds deprecation warnings as requested.

neutrinoceros commented:
pre-commit.ci autofix

neutrinoceros commented:
(rebased again now that #288 went in)

ngoldbaum commented on unyt/tests/test_unyt_array.py:
```diff
@@ -39,6 +31,15 @@
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file COPYING.txt, distributed with this software.
+# ----------------------------------------------------------------------------
```
Can you move this header back to the top of the file?

neutrinoceros replied:
oh, good catch. I think this was done by `isort` or something, but I should have seen it

ngoldbaum commented on unyt/_deprecation.py:
```diff
@@ -0,0 +1,26 @@
+
+
+class UnytDeprecationWarning(UserWarning):
+    pass
```
Can we use the built-in `DeprecationWarning` instead? We already use that elsewhere.

ngoldbaum reviewed (commented):
Just a couple minor nits from the changes you made today. Will merge once they're resolved.
